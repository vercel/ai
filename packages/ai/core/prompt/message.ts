import { z } from 'zod';
import { ProviderMetadata } from '../types';
import {
  providerMetadataSchema,
  ProviderOptions,
} from '../types/provider-metadata';
import {
  FilePart,
  filePartSchema,
  ImagePart,
  imagePartSchema,
  ReasoningPart,
  reasoningPartSchema,
  RedactedReasoningPart,
  redactedReasoningPartSchema,
  TextPart,
  textPartSchema,
  ToolCallPart,
  toolCallPartSchema,
  ToolResultPart,
  toolResultPartSchema,
} from './content-part';

/**
 A system message. It can contain system information.

 Note: using the "system" part of the prompt is strongly preferred
 to increase the resilience against prompt injection attacks,
 and because not all providers support several system messages.
 */
export type CoreSystemMessage = {
  role: 'system';
  content: string;

  /**
Additional provider-specific metadata. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
  providerOptions?: ProviderOptions;

  /**
@deprecated Use `providerOptions` instead.
 */
  experimental_providerMetadata?: ProviderMetadata;
};

export const coreSystemMessageSchema: z.ZodType<CoreSystemMessage> = z.object({
  role: z.literal('system'),
  content: z.string(),
  providerOptions: providerMetadataSchema.optional(),
  experimental_providerMetadata: providerMetadataSchema.optional(),
});

/**
A user message. It can contain text or a combination of text and images.
 */
export type CoreUserMessage = {
  role: 'user';
  content: UserContent;

  /**
Additional provider-specific metadata. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
  providerOptions?: ProviderOptions;

  /**
@deprecated Use `providerOptions` instead.
*/
  experimental_providerMetadata?: ProviderMetadata;
};

export const coreUserMessageSchema: z.ZodType<CoreUserMessage> = z.object({
  role: z.literal('user'),
  content: z.union([
    z.string(),
    z.array(z.union([textPartSchema, imagePartSchema, filePartSchema])),
  ]),
  providerOptions: providerMetadataSchema.optional(),
  experimental_providerMetadata: providerMetadataSchema.optional(),
});

/**
Content of a user message. It can be a string or an array of text and image parts.
 */
export type UserContent = string | Array<TextPart | ImagePart | FilePart>;

/**
An assistant message. It can contain text, tool calls, or a combination of text and tool calls.
 */
export type CoreAssistantMessage = {
  role: 'assistant';
  content: AssistantContent;

  /**
Additional provider-specific metadata. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
  providerOptions?: ProviderOptions;

  /**
@deprecated Use `providerOptions` instead.
*/
  experimental_providerMetadata?: ProviderMetadata;
};

export const coreAssistantMessageSchema: z.ZodType<CoreAssistantMessage> =
  z.object({
    role: z.literal('assistant'),
    content: z.union([
      z.string(),
      z.array(
        z.union([
          textPartSchema,
          filePartSchema,
          reasoningPartSchema,
          redactedReasoningPartSchema,
          toolCallPartSchema,
        ]),
      ),
    ]),
    providerOptions: providerMetadataSchema.optional(),
    experimental_providerMetadata: providerMetadataSchema.optional(),
  });

/**
Content of an assistant message.
It can be a string or an array of text, image, reasoning, redacted reasoning, and tool call parts.
 */
export type AssistantContent =
  | string
  | Array<
      TextPart | FilePart | ReasoningPart | RedactedReasoningPart | ToolCallPart
    >;

/**
A tool message. It contains the result of one or more tool calls.
 */
export type CoreToolMessage = {
  role: 'tool';
  content: ToolContent;

  /**
Additional provider-specific metadata. They are passed through
to the provider from the AI SDK and enable provider-specific
functionality that can be fully encapsulated in the provider.
 */
  providerOptions?: ProviderOptions;

  /**
@deprecated Use `providerOptions` instead.
*/
  experimental_providerMetadata?: ProviderMetadata;
};

export const coreToolMessageSchema: z.ZodType<CoreToolMessage> = z.object({
  role: z.literal('tool'),
  content: z.array(toolResultPartSchema),
  providerOptions: providerMetadataSchema.optional(),
  experimental_providerMetadata: providerMetadataSchema.optional(),
});

/**
Content of a tool message. It is an array of tool result parts.
 */
export type ToolContent = Array<ToolResultPart>;

/**
A message that can be used in the `messages` field of a prompt.
It can be a user message, an assistant message, or a tool message.
 */
export type CoreMessage =
  | CoreSystemMessage
  | CoreUserMessage
  | CoreAssistantMessage
  | CoreToolMessage;

export const coreMessageSchema: z.ZodType<CoreMessage> = z.union([
  coreSystemMessageSchema,
  coreUserMessageSchema,
  coreAssistantMessageSchema,
  coreToolMessageSchema,
]);
