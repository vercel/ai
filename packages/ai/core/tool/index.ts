export { createMCPClient as experimental_createMCPClient } from './mcp/mcp-client';
export type { MCPTransport } from './mcp/mcp-transport';
export { tool } from './tool';
export type { CoreTool, Tool, ToolExecutionOptions } from './tool';
