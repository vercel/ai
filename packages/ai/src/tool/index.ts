export type {
  JSONRPCError,
  JSONRPCMessage,
  JSONRPCNotification,
  JSONRPCRequest,
  JSONRPCResponse,
} from './mcp/json-rpc-message';
export {
  createMCPClient as experimental_createMCPClient,
  type MCPClientConfig as experimental_MCPClientConfig,
  type MCPClient as experimental_MCPClient,
} from './mcp/mcp-client';
export type { MCPTransport } from './mcp/mcp-transport';
