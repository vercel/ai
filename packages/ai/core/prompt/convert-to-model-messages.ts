import { ToolSet } from '../generate-text/tool-set';
import {
  FileUIPart,
  ReasoningUIPart,
  TextUIPart,
  ToolInvocationUIPart,
  UIMessage,
} from '../types';
import { getUIText } from '../ui/get-ui-text';
import { ToolResultPart } from './content-part';
import { AssistantContent, ModelMessage } from './message';
import { MessageConversionError } from './message-conversion-error';

/**
Converts an array of messages from useChat into an array of CoreMessages that can be used
with the AI core functions (e.g. `streamText`).
 */
export function convertToModelMessages<TOOLS extends ToolSet = never>(
  messages: Array<Omit<UIMessage, 'id'>>,
  options?: { tools?: TOOLS },
): ModelMessage[] {
  const tools = options?.tools ?? ({} as TOOLS);
  const modelMessages: ModelMessage[] = [];

  for (const message of messages) {
    switch (message.role) {
      case 'system': {
        modelMessages.push({
          role: 'system',
          content: getUIText(message.parts),
        });
        break;
      }

      case 'user': {
        modelMessages.push({
          role: 'user',
          content: message.parts
            .filter(
              (part): part is TextUIPart | FileUIPart =>
                part.type === 'text' || part.type === 'file',
            )
            .map(part =>
              part.type === 'file'
                ? {
                    type: 'file' as const,
                    mediaType: part.mediaType,
                    filename: part.filename,
                    data: part.url,
                  }
                : part,
            ),
        });

        break;
      }

      case 'assistant': {
        if (message.parts != null) {
          let currentStep = 0;
          let blockHasToolInvocations = false;
          let block: Array<
            TextUIPart | ToolInvocationUIPart | ReasoningUIPart | FileUIPart
          > = [];

          function processBlock() {
            const content: AssistantContent = [];

            for (const part of block) {
              switch (part.type) {
                case 'text': {
                  content.push(part);
                  break;
                }
                case 'file': {
                  content.push({
                    type: 'file' as const,
                    mediaType: part.mediaType,
                    data: part.url,
                  });
                  break;
                }
                case 'reasoning': {
                  content.push({
                    type: 'reasoning' as const,
                    text: part.text,
                    providerOptions: part.providerMetadata,
                  });
                  break;
                }
                case 'tool-invocation':
                  content.push({
                    type: 'tool-call' as const,
                    toolCallId: part.toolInvocation.toolCallId,
                    toolName: part.toolInvocation.toolName,
                    args: part.toolInvocation.args,
                  });
                  break;
                default: {
                  const _exhaustiveCheck: never = part;
                  throw new Error(`Unsupported part: ${_exhaustiveCheck}`);
                }
              }
            }

            modelMessages.push({
              role: 'assistant',
              content,
            });

            // check if there are tool invocations with results in the block
            const stepInvocations = block
              .filter(
                (
                  part:
                    | TextUIPart
                    | ToolInvocationUIPart
                    | ReasoningUIPart
                    | FileUIPart,
                ): part is ToolInvocationUIPart =>
                  part.type === 'tool-invocation',
              )
              .map(part => part.toolInvocation);

            // tool message with tool results
            if (stepInvocations.length > 0) {
              modelMessages.push({
                role: 'tool',
                content: stepInvocations.map(
                  (toolInvocation): ToolResultPart => {
                    if (!('result' in toolInvocation)) {
                      throw new MessageConversionError({
                        originalMessage: message,
                        message:
                          'ToolInvocation must have a result: ' +
                          JSON.stringify(toolInvocation),
                      });
                    }

                    const { toolCallId, toolName, result } = toolInvocation;

                    const tool = tools[toolName];
                    return tool?.experimental_toToolResultContent != null
                      ? {
                          type: 'tool-result',
                          toolCallId,
                          toolName,
                          result: tool.experimental_toToolResultContent(result),
                          experimental_content:
                            tool.experimental_toToolResultContent(result),
                        }
                      : {
                          type: 'tool-result',
                          toolCallId,
                          toolName,
                          result,
                        };
                  },
                ),
              });
            }

            // updates for next block
            block = [];
            blockHasToolInvocations = false;
            currentStep++;
          }

          for (const part of message.parts) {
            switch (part.type) {
              case 'text': {
                if (blockHasToolInvocations) {
                  processBlock(); // text must come before tool invocations
                }
                block.push(part);
                break;
              }
              case 'file':
              case 'reasoning': {
                block.push(part);
                break;
              }
              case 'tool-invocation': {
                if ((part.toolInvocation.step ?? 0) !== currentStep) {
                  processBlock();
                }
                block.push(part);
                blockHasToolInvocations = true;
                break;
              }
            }
          }

          processBlock();

          break;
        }

        break;
      }

      default: {
        const _exhaustiveCheck: never = message.role;
        throw new MessageConversionError({
          originalMessage: message,
          message: `Unsupported role: ${_exhaustiveCheck}`,
        });
      }
    }
  }

  return modelMessages;
}

/**
@deprecated Use `convertToModelMessages` instead.
 */
// TODO remove in AI SDK 6
export const convertToCoreMessages = convertToModelMessages;
