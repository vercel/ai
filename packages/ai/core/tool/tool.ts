import { Schema } from '@ai-sdk/ui-utils';
import { z } from 'zod';
import { ToolResultContent } from '../prompt/tool-result-content';
import { CoreMessage } from '../prompt/message';

export type ToolParameters = z.ZodTypeAny | Schema<any>;

export type inferParameters<PARAMETERS extends ToolParameters> =
  PARAMETERS extends Schema<any>
    ? PARAMETERS['_type']
    : PARAMETERS extends z.ZodTypeAny
    ? z.infer<PARAMETERS>
    : never;

export interface ToolExecutionOptions {
  /**
   * The ID of the tool call. You can use it e.g. when sending tool-call related information with stream data.
   */
  toolCallId: string;

  /**
   * Messages that were sent to the language model to initiate the response that contained the tool call.
   * The messages **do not** include the system prompt nor the assistant response that contained the tool call.
   */
  messages: CoreMessage[];

  /**
   * An optional abort signal that indicates that the overall operation should be aborted.
   */
  abortSignal?: AbortSignal;
}

/**
A tool contains the description and the schema of the input that the tool expects.
This enables the language model to generate the input.

The tool can also contain an optional execute function for the actual execution function of the tool.
 */
export type Tool<PARAMETERS extends ToolParameters = any, RESULT = any> = {
  /**
The schema of the input that the tool expects. The language model will use this to generate the input.
It is also used to validate the output of the language model.
Use descriptions to make the input understandable for the language model.
   */
  parameters: PARAMETERS;

  /**
An optional description of what the tool does.
Will be used by the language model to decide whether to use the tool.
Not used for provider-defined tools.
   */
  description?: string;

  /**
Optional conversion function that maps the tool result to multi-part tool content for LLMs.
   */
  experimental_toToolResultContent?: (result: RESULT) => ToolResultContent;

  /**
An async function that is called with the arguments from the tool call and produces a result.
If not provided, the tool will not be executed automatically.

@args is the input of the tool call.
@options.abortSignal is a signal that can be used to abort the tool call.
   */
  execute?: (
    args: inferParameters<PARAMETERS>,
    options: ToolExecutionOptions,
  ) => PromiseLike<RESULT>;
} & (
  | {
      /**
Function tool.
       */
      type?: undefined | 'function';
    }
  | {
      /**
Provider-defined tool.
       */
      type: 'provider-defined';

      /**
The ID of the tool. Should follow the format `<provider-name>.<tool-name>`.
       */
      id: `${string}.${string}`;

      /**
The arguments for configuring the tool. Must match the expected arguments defined by the provider for this tool.
       */
      args: Record<string, unknown>;
    }
);

/**
 * @deprecated Use `Tool` instead.
 */
// TODO remove in v5
export type CoreTool<
  PARAMETERS extends ToolParameters = any,
  RESULT = any,
> = Tool<PARAMETERS, RESULT>;

/**
Helper function for inferring the execute args of a tool.
 */
// Note: special type inference is needed for the execute function args to make sure they are inferred correctly.
export function tool<PARAMETERS extends ToolParameters, RESULT>(
  tool: Tool<PARAMETERS, RESULT> & {
    execute: (
      args: inferParameters<PARAMETERS>,
      options: ToolExecutionOptions,
    ) => PromiseLike<RESULT>;
  },
): Tool<PARAMETERS, RESULT> & {
  execute: (
    args: inferParameters<PARAMETERS>,
    options: ToolExecutionOptions,
  ) => PromiseLike<RESULT>;
};
export function tool<PARAMETERS extends ToolParameters, RESULT>(
  tool: Tool<PARAMETERS, RESULT> & {
    execute?: undefined;
  },
): Tool<PARAMETERS, RESULT> & {
  execute: undefined;
};
export function tool<PARAMETERS extends ToolParameters, RESULT = any>(
  tool: Tool<PARAMETERS, RESULT>,
): Tool<PARAMETERS, RESULT> {
  return tool;
}
