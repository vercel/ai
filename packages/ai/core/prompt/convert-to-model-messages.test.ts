import { z } from 'zod';
import { tool } from '../tool/tool';
import { convertToModelMessages } from './convert-to-model-messages';
import { ModelMessage } from './message';

describe('convertToModelMessages', () => {
  describe('system message', () => {
    it('should convert a simple system message', () => {
      const result = convertToModelMessages([
        {
          role: 'system',
          content: 'System message',
          parts: [{ text: 'System message', type: 'text' }],
        },
      ]);

      expect(result).toEqual([{ role: 'system', content: 'System message' }]);
    });
  });

  describe('user message', () => {
    it('should convert a simple user message', () => {
      const result = convertToModelMessages([
        {
          role: 'user',
          content: 'Hello, AI!',
          parts: [{ text: 'Hello, AI!', type: 'text' }],
        },
      ]);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "Hello, AI!",
                "type": "text",
              },
            ],
            "role": "user",
          },
        ]
      `);
    });

    it('should prefer content in parts when content is empty', () => {
      const result = convertToModelMessages([
        {
          role: 'user',
          content: '', // empty content
          parts: [
            {
              type: 'text',
              text: 'hey, how is it going?',
            },
          ],
        },
      ]);

      expect(result).toEqual([
        {
          role: 'user',
          content: [{ type: 'text', text: 'hey, how is it going?' }],
        },
      ]);
    });

    it('should handle user message file parts', () => {
      const result = convertToModelMessages([
        {
          role: 'user',
          content: 'Check this image',
          parts: [
            {
              type: 'file',
              mediaType: 'image/jpeg',
              url: 'https://example.com/image.jpg',
            },
            { type: 'text', text: 'Check this image' },
          ],
        },
      ]);

      expect(result).toEqual([
        {
          role: 'user',
          content: [
            {
              type: 'file',
              mediaType: 'image/jpeg',
              data: 'https://example.com/image.jpg',
            },
            { type: 'text', text: 'Check this image' },
          ],
        },
      ]);
    });
  });

  describe('assistant message', () => {
    it('should convert a simple assistant message', () => {
      const result = convertToModelMessages([
        {
          role: 'assistant',
          content: '', // empty content
          parts: [{ type: 'text', text: 'Hello, human!' }],
        },
      ]);

      expect(result).toEqual([
        {
          role: 'assistant',
          content: [{ type: 'text', text: 'Hello, human!' }],
        },
      ]);
    });

    it('should convert an assistant message with reasoning', () => {
      const result = convertToModelMessages([
        {
          role: 'assistant',
          content: '', // empty content
          parts: [
            {
              type: 'reasoning',
              text: 'Thinking...',
              providerMetadata: {
                testProvider: {
                  signature: '1234567890',
                },
              },
            },
            {
              type: 'reasoning',
              text: 'redacted-data',
              providerMetadata: {
                testProvider: { isRedacted: true },
              },
            },
            { type: 'text', text: 'Hello, human!' },
          ],
        },
      ]);

      expect(result).toEqual([
        {
          role: 'assistant',
          content: [
            {
              type: 'reasoning',
              text: 'Thinking...',
              providerOptions: { testProvider: { signature: '1234567890' } },
            },
            {
              type: 'reasoning',
              text: 'redacted-data',
              providerOptions: { testProvider: { isRedacted: true } },
            },
            { type: 'text', text: 'Hello, human!' },
          ],
        },
      ] satisfies ModelMessage[]);
    });

    it('should convert an assistant message with file parts', () => {
      const result = convertToModelMessages([
        {
          role: 'assistant',
          content: '', // empty content
          parts: [
            {
              type: 'file',
              mediaType: 'image/png',
              url: 'data:image/png;base64,dGVzdA==',
            },
          ],
        },
      ]);

      expect(result).toEqual([
        {
          role: 'assistant',
          content: [
            {
              type: 'file',
              mediaType: 'image/png',
              data: 'data:image/png;base64,dGVzdA==',
            },
          ],
        },
      ] satisfies ModelMessage[]);
    });

    it('should handle assistant message with tool invocations', () => {
      const result = convertToModelMessages([
        {
          role: 'assistant',
          content: '', // empty content
          parts: [
            { type: 'text', text: 'Let me calculate that for you.' },
            {
              type: 'tool-invocation',
              toolInvocation: {
                state: 'result',
                toolCallId: 'call1',
                toolName: 'calculator',
                args: { operation: 'add', numbers: [1, 2] },
                result: '3',
                step: 0,
              },
            },
          ],
        },
      ]);

      expect(result).toMatchSnapshot();
    });

    it('should handle assistant message with tool invocations that have multi-part responses', () => {
      const tools = {
        screenshot: tool({
          parameters: z.object({}),
          execute: async () => 'imgbase64',
          experimental_toToolResultContent: result => [
            { type: 'image', data: result },
          ],
        }),
      };

      const result = convertToModelMessages(
        [
          {
            role: 'assistant',
            content: '', // empty content
            parts: [
              { type: 'text', text: 'Let me calculate that for you.' },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call1',
                  toolName: 'screenshot',
                  args: {},
                  result: 'imgbase64',
                  step: 0,
                },
              },
            ],
          },
        ],
        { tools }, // separate tools to ensure that types are inferred correctly
      );

      expect(result).toMatchSnapshot();
    });

    it('should handle conversation with an assistant message that has empty tool invocations', () => {
      const result = convertToModelMessages([
        {
          role: 'user',
          content: 'text1',
          parts: [{ type: 'text', text: 'text1' }],
        },
        {
          role: 'assistant',
          content: '', // empty content
          parts: [{ type: 'text', text: 'text2' }],
        },
      ]);

      expect(result).toMatchSnapshot();
    });

    it('should handle conversation with multiple tool invocations that have step information', () => {
      const tools = {
        screenshot: tool({
          parameters: z.object({ value: z.string() }),
          execute: async () => 'imgbase64',
        }),
      };

      const result = convertToModelMessages(
        [
          {
            role: 'assistant',
            content: '', // empty content
            parts: [
              { type: 'text', text: 'response' },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-1',
                  toolName: 'screenshot',
                  args: { value: 'value-1' },
                  result: 'result-1',
                  step: 0,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-2',
                  toolName: 'screenshot',
                  args: { value: 'value-2' },
                  result: 'result-2',
                  step: 1,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-3',
                  toolName: 'screenshot',
                  args: { value: 'value-3' },
                  result: 'result-3',
                  step: 1,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-4',
                  toolName: 'screenshot',
                  args: { value: 'value-4' },
                  result: 'result-4',
                  step: 2,
                },
              },
            ],
          },
        ],
        { tools }, // separate tools to ensure that types are inferred correctly
      );

      expect(result).toMatchSnapshot();
    });

    it('should handle conversation with mix of tool invocations and text', () => {
      const tools = {
        screenshot: tool({
          parameters: z.object({ value: z.string() }),
          execute: async () => 'imgbase64',
        }),
      };

      const result = convertToModelMessages(
        [
          {
            role: 'assistant',
            content: '', // empty content
            parts: [
              { type: 'text', text: 'i am gonna use tool1' },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-1',
                  toolName: 'screenshot',
                  args: { value: 'value-1' },
                  result: 'result-1',
                  step: 0,
                },
              },
              { type: 'text', text: 'i am gonna use tool2 and tool3' },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-2',
                  toolName: 'screenshot',
                  args: { value: 'value-2' },
                  result: 'result-2',
                  step: 1,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-3',
                  toolName: 'screenshot',
                  args: { value: 'value-3' },
                  result: 'result-3',
                  step: 1,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-4',
                  toolName: 'screenshot',
                  args: { value: 'value-4' },
                  result: 'result-4',
                  step: 2,
                },
              },
              { type: 'text', text: 'final response' },
            ],
          },
        ],
        { tools }, // separate tools to ensure that types are inferred correctly
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('multiple messages', () => {
    it('should handle a conversation with multiple messages', () => {
      const result = convertToModelMessages([
        {
          role: 'user',
          content: "What's the weather like?",
          parts: [{ type: 'text', text: "What's the weather like?" }],
        },
        {
          role: 'assistant',
          content: '',
          parts: [{ type: 'text', text: "I'll check that for you." }],
        },
        {
          role: 'user',
          content: 'Thanks!',
          parts: [{ type: 'text', text: 'Thanks!' }],
        },
      ]);

      expect(result).toMatchInlineSnapshot(`
        [
          {
            "content": [
              {
                "text": "What's the weather like?",
                "type": "text",
              },
            ],
            "role": "user",
          },
          {
            "content": [
              {
                "text": "I'll check that for you.",
                "type": "text",
              },
            ],
            "role": "assistant",
          },
          {
            "content": [
              {
                "text": "Thanks!",
                "type": "text",
              },
            ],
            "role": "user",
          },
        ]
      `);
    });

    it('should handle conversation with multiple tool invocations and user message at the end', () => {
      const tools = {
        screenshot: tool({
          parameters: z.object({ value: z.string() }),
          execute: async () => 'imgbase64',
        }),
      };

      const result = convertToModelMessages(
        [
          {
            role: 'assistant',
            content: '',
            parts: [
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-1',
                  toolName: 'screenshot',
                  args: { value: 'value-1' },
                  result: 'result-1',
                  step: 0,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-2',
                  toolName: 'screenshot',
                  args: { value: 'value-2' },
                  result: 'result-2',
                  step: 1,
                },
              },

              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-3',
                  toolName: 'screenshot',
                  args: { value: 'value-3' },
                  result: 'result-3',
                  step: 1,
                },
              },
              {
                type: 'tool-invocation',
                toolInvocation: {
                  state: 'result',
                  toolCallId: 'call-4',
                  toolName: 'screenshot',
                  args: { value: 'value-4' },
                  result: 'result-4',
                  step: 2,
                },
              },
              { type: 'text', text: 'response' },
            ],
          },
          {
            role: 'user',
            content: 'Thanks!',
            parts: [{ type: 'text', text: 'Thanks!' }],
          },
        ],
        { tools }, // separate tools to ensure that types are inferred correctly
      );

      expect(result).toMatchSnapshot();
    });
  });

  describe('error handling', () => {
    it('should throw an error for unhandled roles', () => {
      expect(() => {
        convertToModelMessages([
          {
            role: 'unknown' as any,
            content: 'unknown role message',
            parts: [{ text: 'unknown role message', type: 'text' }],
          },
        ]);
      }).toThrow('Unsupported role: unknown');
    });
  });
});
