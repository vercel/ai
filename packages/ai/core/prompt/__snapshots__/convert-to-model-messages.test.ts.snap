// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`convertToModelMessages > assistant message > should handle assistant message with tool invocations 1`] = `
[
  {
    "content": [
      {
        "text": "Let me calculate that for you.",
        "type": "text",
      },
      {
        "args": {
          "numbers": [
            1,
            2,
          ],
          "operation": "add",
        },
        "toolCallId": "call1",
        "toolName": "calculator",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "3",
        "toolCallId": "call1",
        "toolName": "calculator",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
]
`;

exports[`convertToModelMessages > assistant message > should handle assistant message with tool invocations that have multi-part responses 1`] = `
[
  {
    "content": [
      {
        "text": "Let me calculate that for you.",
        "type": "text",
      },
      {
        "args": {},
        "toolCallId": "call1",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "experimental_content": [
          {
            "data": "imgbase64",
            "type": "image",
          },
        ],
        "result": [
          {
            "data": "imgbase64",
            "type": "image",
          },
        ],
        "toolCallId": "call1",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
]
`;

exports[`convertToModelMessages > assistant message > should handle conversation with an assistant message that has empty tool invocations 1`] = `
[
  {
    "content": [
      {
        "text": "text1",
        "type": "text",
      },
    ],
    "role": "user",
  },
  {
    "content": [
      {
        "text": "text2",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
]
`;

exports[`convertToModelMessages > assistant message > should handle conversation with mix of tool invocations and text 1`] = `
[
  {
    "content": [
      {
        "text": "i am gonna use tool1",
        "type": "text",
      },
      {
        "args": {
          "value": "value-1",
        },
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-1",
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "text": "i am gonna use tool2 and tool3",
        "type": "text",
      },
      {
        "args": {
          "value": "value-2",
        },
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-call",
      },
      {
        "args": {
          "value": "value-3",
        },
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-2",
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-result",
      },
      {
        "result": "result-3",
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "args": {
          "value": "value-4",
        },
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-4",
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "text": "final response",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
]
`;

exports[`convertToModelMessages > assistant message > should handle conversation with multiple tool invocations that have step information 1`] = `
[
  {
    "content": [
      {
        "text": "response",
        "type": "text",
      },
      {
        "args": {
          "value": "value-1",
        },
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-1",
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "args": {
          "value": "value-2",
        },
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-call",
      },
      {
        "args": {
          "value": "value-3",
        },
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-2",
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-result",
      },
      {
        "result": "result-3",
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "args": {
          "value": "value-4",
        },
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-4",
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
]
`;

exports[`convertToModelMessages > multiple messages > should handle conversation with multiple tool invocations and user message at the end 1`] = `
[
  {
    "content": [
      {
        "args": {
          "value": "value-1",
        },
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-1",
        "toolCallId": "call-1",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "args": {
          "value": "value-2",
        },
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-call",
      },
      {
        "args": {
          "value": "value-3",
        },
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-2",
        "toolCallId": "call-2",
        "toolName": "screenshot",
        "type": "tool-result",
      },
      {
        "result": "result-3",
        "toolCallId": "call-3",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "args": {
          "value": "value-4",
        },
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-call",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "result": "result-4",
        "toolCallId": "call-4",
        "toolName": "screenshot",
        "type": "tool-result",
      },
    ],
    "role": "tool",
  },
  {
    "content": [
      {
        "text": "response",
        "type": "text",
      },
    ],
    "role": "assistant",
  },
  {
    "content": [
      {
        "text": "Thanks!",
        "type": "text",
      },
    ],
    "role": "user",
  },
]
`;
